solara/components/input.py

```diff
@@ -276,113 +288,70 @@ def Page():
-        if isinstance(internal_value, str):
-            try:
-                numerical_value = ast.literal_eval(internal_value.replace(",", "."))
-            except Exception:
-                set_internal_value(value)
-                # this make sure the current value gets updated directly
-                internal_value = str(value)
-            else:
-                if numerical_value != value:
-                    set_internal_value(str(value))
-                    # this make sure the current value gets updated directly
-                    internal_value = str(value)
-        elif internal_value is not None:
-            if internal_value != value:
-                set_internal_value(value)
-                # this make sure the current value gets updated directly
-                internal_value = value
-        elif internal_value is None:
-            set_internal_value(value)
-            # this make sure the current value gets updated directly
-            internal_value = value
-
-    # make sure that if the external value changes, our internal model gets updated
-    # but if the internal model is in a different representation
-    # (e.g. internal_value='1e2' and internal_value=100) we don't want to change
-    # our internal model
-    solara.use_memo(on_external_value_change, [reactive_value.value])
-
-    def internal_value_check_type():
-        if isinstance(internal_value, str):
-            try:
-                numerical_proper_type = str_to_numeric(parse(internal_value))
-                numerical = parse(internal_value)
-            except Exception:
-                return internal_value
-            else:
-                if numerical_proper_type != numerical:
-                    return str(numerical_proper_type)
-                else:
-                    return internal_value
-        else:
-            return internal_value
+    def hook():
+        def on_external_value_change(new_value: Optional[T]):
+            new_string_value = stringify(new_value)
+            if new_value != parse(string_value_ref.current):
+                set_string_value(new_string_value)
 
-    # make sure that when internal_value="1.1", but str_to_numeric=int
-    # internal value becomes "1"
-    internal_value = solara.use_memo(internal_value_check_type, [internal_value])
+        return reactive_value.subscribe(on_external_value_change)
 
-    def set_value_cast(value):
-        set_internal_value(value)
-        if value is None or value == "":
-            if optional:
-                set_error(False)
-                reactive_value.set(None)  # type: ignore
-            else:
-                set_error(True)
-            return
-        try:
-            numeric_value = str_to_numeric(parse(value))
-        except Exception:
-            set_error(True)
-        else:
-            set_error(False)
-            reactive_value.set(numeric_value)
+    solara.use_effect(hook, [reactive_value.value])
```

Comment:
```suggestion
    solara.use_effect(sync_back_input_value, [reactive_value])
```
This this only re-attach when the reactive_value object changes (very uncommon edge case that can happen when the parent component sometimes passes a reactive var, and sometime a normal var). It doesn't need to re-attach when the value changes.